### PR TITLE
Add aspect ratio presets support via theme.json

### PR DESCRIPTION
**Pull request by jffng**
## What?
This PR adds support for aspect ratio presets via theme.json. It adds by default the same presets that are available for the crop utilities in the image block.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/38990

## How?
Bootstraps the presets via the theme.json class.

## Testing Instructions
- Check out this PR and verify you see the presets in the following places:
    - Core Image Block Inspector Controls
    - Core Image Block Croping tools (Toolbar)
    - Core Featured Image Inspector Controls
    - Core Cover block Inspector Controls
- Add your own presets via the theme (`settings.dimensions.aspectRatios`) and verify they appear
- Disable the default presets `settings.dimensions.defaultAspectRatios` -> `false` and verify they do not appear

<details>
<summary> Sample `theme.json`</summary>

```json
{
	"version": 3,
	"settings": {
		"dimensions": {
			"defaultAspectRatios": true,
			"aspectRatios": [
				{
					"name": "Extra Wide - 2:1",
					"ratio": "2/1",
					"slug": "2-1"
				}
			]
		}
	}
}

```

</details>

### Testing Instructions for Keyboard

N/A, backend only change. 

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/20684594/9050f661-1452-4699-8822-93b2fd52466c

### CODE REVIEW

**Comment by mtias** [comment]
Could we just go with `ratio` and avoid `name` and `slug`?

**Comment by jffng** [comment]
@mtias I believe `slug` is currently [required by the Style Engine](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/class-wp-style-engine.php#L33) in order to generate the preset. 

We can drop `name` if that's not helpful, but I was not sure if there are cases where we'd want to name a ratio differently, e.g. `Square` for `1/1`.

**Comment by Mamaduka** [comment]
Hi, @jffng

I just wanted to check the status of the feature. Are we still aiming for WP 6.2?

**Comment by ajlende** [comment]
I have some bandwidth to work on this right now if @jffng is busy with other things.

File: lib/class-wp-theme-json-gutenberg.php
Comment (ajlende) [inline lib/class-wp-theme-json-gutenberg.php:138]: I wanted to add just the settings for the image block aspect ratio tool in this PR, but it seems that the UI is coupled to the preset generation. So, even though this generated CSS isn't used, it had to be added to get the UI to work.

I don't think it's too big of a deal since we would add those presets for the `aspect-ratio` property anyway for the featured image in a follow-up, but it certainly surprised me that the code was coupled this way.

File: lib/theme.json
Comment (ajlende) [inline lib/theme.json:195]: These default presets are currently unused as well so the UI can continue to be grouped by portrait and landscape to match the design. It's unfortunate that [we aren't using JSON with comments](https://github.com/WordPress/gutenberg/issues/35099), so I can't really comment in the code here about it.

Comment (ajlende) [reply lib/theme.json:195]: I think I'm going to update the UI to use these and add a special case for the defaults to categorize them as landscape or portrait.

We could also categorize the theme supplied ones as well if we update the designs. Or we could choose to not categorize the defaults to be consistent with how other settings presets work. @richtabor maybe you have thoughts?

Comment (richtabor) [reply lib/theme.json:195]: > Or we could choose to not categorize the defaults to be consistent with how other settings presets work 

You mean keep the "Theme" section, with the theme's defaults, v. putting those defaults within the proper landscape/portrait sections?

Comment (ajlende) [reply lib/theme.json:195]: The theme ones would be as-is and the landscape and portrait get grouped into just default. For example, in the color/gradient presets, they are grouped by default and theme. And in global styles, there's an additional custom section.

global styles color presets

Comment (richtabor) [reply lib/theme.json:195]: > The theme ones would be as-is and the landscape and portrait get grouped into just default. For example, in the color/gradient presets, they are grouped by default and theme. 

Would we loose the separation of portrait and landscape? 

Is it a heavy lift to get theme presets in? If so, I'm ok dropping support for theme-specific presets (for now) and circle back around for that.

Comment (ajlende) [reply lib/theme.json:195]: I can organize it however you'd like. I pushed up a change to include the defaults from the default theme.json categorized under landscape and portrait already. It also isn't any more difficult to change how they're categorized if you think it makes sense to do it differently.

Comment (richtabor) [reply lib/theme.json:195]: Ok cool, good deal.

**Comment by ajlende** [comment]
I've been working on creating a new select component that allows for dividers and headings to match the designs. I can't reuse the existing dropdown because it's a toolbar dropdown, not an input/select dropdown. And I can't use the existing SelectControl or CustomSelectControl because they don't support custom children like our section headers or dividers.

I figured I could reuse the existing presentational components of the CustomSelectControl, but they come with [too much baggage](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/custom-select-control/index.js): backwards compatibility, future compatibility, experimental options and features that the new component wouldn't use. It also seems bad to not use those presentational components since they should, in theory, be the right thing to use.

So in the interest of getting this shipped and iterated on, I'm going to deviate from the designs and use the existing SelectControl with the "current aspect" option at top followed by the theme options and then the default options without any of the separators or headers.

**Comment by richtabor** [comment]
> I'm going to deviate from the designs and use the existing SelectControl with the "current aspect" option at top followed by the theme options and then the default options without any of the separators or headers.

That works. We can iterate towards it. 

>  followed by the theme options 

After talking with @mtias, it would better to hold off on the theme defaults aspect of this — to give us time to ensure the theme aspect ratios preset spec is clear/expected/well-defined.

**Comment by ndiego** [comment]
Hey @richtabor @ajlende, this PR is on the [6.3 project board](https://github.com/orgs/WordPress/projects/103/views/1). It's been a while since this was worked on. Is this still something we are aiming to get into 6.3?

**Comment by richtabor** [comment]
> Is this still something we are aiming to get into 6.3?

Not not currently; we can remove.

**Comment by unscripted** [comment]
I would love to see this in an upcoming release. Unless I'm missing it, we cannot add a custom aspect ratio via theme.json.

**Comment by strarsis** [comment]
+1, this would be such a great addition! Being able to define the crop interactively for each image without extra plugins.

**Comment by markhowellsmead** [comment]
Bumping this to hopefully get some traction for filterable or configurable custom aspect ratios.

**Comment by fabiankaegy** [comment]
I just rebased this PR in order to make it easier to test and evolve the API :)

File: packages/block-editor/src/components/aspect-ratio-dropdown/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/aspect-ratio-dropdown/index.js]: Let's combine the setting getters - `useSettings( 'dimensions.aspectRatios.default', 'dimensions.aspectRatios.theme', 'dimensions.defaultAspectRatios' )`

Related: https://make.wordpress.org/core/2024/03/09/miscellaneous-editor-changes-in-wordpress-6-5/#new-usesettings-hook-for-reading-block-instance-settings

File: packages/block-editor/src/components/aspect-ratio-dropdown/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/aspect-ratio-dropdown/index.js]: Having unit tests for these utils would be helpful.

File: packages/block-editor/src/components/aspect-ratio-dropdown/index.js
Comment (fabiankaegy) [reply packages/block-editor/src/components/aspect-ratio-dropdown/index.js]: Ahh very good call 🤦

Comment (fabiankaegy) [reply packages/block-editor/src/components/aspect-ratio-dropdown/index.js]: Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/001391b1a09287d798116a7e9fa1b29c05e4dffe

File: packages/block-editor/src/components/aspect-ratio-dropdown/index.js
Comment (fabiankaegy) [reply packages/block-editor/src/components/aspect-ratio-dropdown/index.js]: Added a test in https://github.com/WordPress/gutenberg/pull/47271/commits/18c204517ea7f6f82b416f2e56f00355216fcbb8

File: packages/block-editor/src/components/image-editor/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/image-editor/index.js]: What was the reason for extracting this in a separate component? It seems to be used only by the image editor.

File: packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js
Comment (Mamaduka) [inline packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: Unlike `trunk`, this will merge provided options instead of overriding the defaults.

Comment (fabiankaegy) [reply packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: Yes and no. The defaults are now no longer hardcoded here in this JS file but instead are now coming from the core `theme.json` file. 

So the `showDefaultRatios` check determines if they display or not. The `options` currently were not used in the Gutenberg codebase at all. I now needed to use them in the unit tests since the global styles are not populated there. But I'm not sure this is breaking the expected behavior 🤔

File: packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js
Comment (Mamaduka) [inline packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: Instead of merging `options` above, we can pass `options ?? aspectRatioOptions` and preserve the previous component behavior.

Comment (fabiankaegy) [reply packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: That works :)

Comment (fabiankaegy) [reply packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/7f6223fe1c8f24eeed53b9e83a5697006201587e :)

File: packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js
Comment (fabiankaegy) [reply packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js]: Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/7f6223fe1c8f24eeed53b9e83a5697006201587e

File: packages/block-editor/src/components/image-editor/index.js
Comment (fabiankaegy) [reply packages/block-editor/src/components/image-editor/index.js]: That is a very good question. This was already done before I took over the branch. 🤔

Comment (Mamaduka) [reply packages/block-editor/src/components/image-editor/index.js]: One reason I think we should group it with the Image Editor is that the dropdown also resizes the image on the server. Having it as a stand-alone component could be misleading.

Comment (fabiankaegy) [reply packages/block-editor/src/components/image-editor/index.js]: That's a great reason to keep it together :) moved it in https://github.com/WordPress/gutenberg/pull/47271/commits/8a3088c6f56223871f1c8c3dd193791ad4f48944

File: lib/class-wp-theme-json-gutenberg.php
Comment (Mamaduka) [inline lib/class-wp-theme-json-gutenberg.php:410]: Unrelated: I noticed that you can't disable aspect ration for Image block using this setting. cc @draganescu

**Comment by fabiankaegy** [comment]
@ajlende wanted to tag you specifically here for feedback since @oandregal mentioned you recently worked on this area and also you were one of the earlier contributors to this PR before I took it over :)

**Comment by bacoords** [comment]
Did some user testing on this branch and was able to get through all of the instructions testing custom & default aspect ratios across multiple block themes.

File: lib/class-wp-theme-json-gutenberg.php
Comment (draganescu) [reply lib/class-wp-theme-json-gutenberg.php:410]: Was it supposed to be 😄 What is the expected behaviour here:

- does setting dimensions ≻ aspect ratio ≻ null disable the aspect ratio tools somewhere else?
- should it disappear from the UI of the image block? It does not appear in styles, but in settings, whereas for the cover block the UI appears in styles not in settings. What's preffered?
- should it be null or false? When I set theme.json ≻ settings ≻ dimensions ≻ aspectRatio the editor says it should be bool

Comment (Mamaduka) [reply lib/class-wp-theme-json-gutenberg.php:410]: The initial `null` value is how theme json defines flag settings, and it's not important at the moment.

Setting it to `false` disables the aspect ratio UI for blocks. You can test it with the Cover block. I think behavior should be consistent across the blocks that use aspect ratio control.

cc @andrewserong

Comment (draganescu) [reply lib/class-wp-theme-json-gutenberg.php:410]: Yes, indeed setting to false hides this dimension tool from the cover block but it does not hide it from the image block. The reason is that the cover block uses the supports while the image block plops the control in the inspector.

Comment (Mamaduka) [reply lib/class-wp-theme-json-gutenberg.php:410]: Yes. I'm aware of that. Still, the behavior should be consistent.

@fabiankaegy, do you mind creating an issue for the Image and Featured Image block follow-ups we discussed in Slack?

**Comment by bacoords** [comment]
One note about the testing instructions. Should it list the specific blocks that use aspect ratios? Here's what I tested but I may have missed some: 

- Image 
- Cover
- Featured Image

File: lib/class-wp-theme-json-gutenberg.php
Comment (draganescu) [reply lib/class-wp-theme-json-gutenberg.php:410]: > Yes. I'm aware of that.

I knew that, you told me where to look! 😂  I mentioned here for archaeology reasons.

**Comment by fabiankaegy** [comment]
> One note about the testing instructions. Should it list the specific blocks that use aspect ratios? Here's what I tested but I may have missed some:
> 
> * Image
> * Cover
> * Featured Image

@bacoords I'll update the testing instructions to be a little more concise. But you are correct those are all the instances. The one thing to note it that for the image block it shows in two places. The sidebar setting for the aspect ratio and then also in the croping tools in the sidebar

**Review by Mamaduka — APPROVED** [review]
Let's merge this. I don't see any blockers here; we can handle feature unification in follow-ups.

File: lib/class-wp-theme-json-gutenberg.php
Comment (fabiankaegy) [reply lib/class-wp-theme-json-gutenberg.php:410]: Created a follow up issue for the aspect ratio block support cleanup: https://github.com/WordPress/gutenberg/issues/61432

**Comment by fabiankaegy** [comment]
I just created a sync PR on WordPress Develop for the PHP updates to the theme.json class https://core.trac.wordpress.org/ticket/61137
https://github.com/WordPress/wordpress-develop/pull/6567

**Comment by swissspidy** [comment]
Internationalization needs to be fixed before this lands in core. See #61767

**Comment by fabiankaegy** [comment]
@swissspidy thanks for calling this out. I have created #61774 to fix this here in Gutenberg and also updated my PR in core to include the fix. 

Question for you. These strings were previously already present in core. Just hardcoded in JS and piped through `__()`. Will these existing translations get mapped to the new strings automatically or will they need to get translated from scratch?

**Comment by swissspidy** [comment]
They will be marked as fuzzy because now they have proper context (_x vs __)

**Comment by fabiankaegy** [comment]
## Dev Note Draft 


# Add custom Aspect Ratio presets through `theme.json`

WordPress 6.6 adds a new two new properties to the `settings.dimensions` object in `theme.json`. `settings.dimensions.aspectRatios` allows defining your own custom Aspect Ratio presets. These presets will be used by any block that uses the `aspectRatio` block support. In core that means the Image, Featured Image, and Cover block. 

```json
{
	"version": 2,
	"settings": {
		"dimensions": {
			"aspectRatios": [
				{
 					"name": "Square - 1:1",
 					"slug": "square",
 					"ratio": "1"
 				},
 				{
 					"name": "Standard - 4:3",
 					"slug": "4-3",
 					"ratio": "4/3"
 				},
			]
		}
	}
}
```

Additionally the `settings.dimensions.defaultAspectRatios` key allows you to disable the list of core aspect ratio presets.

```json
{
	"version": 2,
	"settings": {
		"dimensions": {
			"defaultAspectRatios": false
		}
	}
}
```

**Pull request merged**
